Add a "Coq Configuration Basics" section

jfehrle opened the pull request:
Adds a "Coq Configuration Basics" section written with more of a tutorial flavor compared to other parts of the doc.  It's meant to complement existing material on configuration (much of which could be improved BTW).  I believe most users want to minimize the time they spend learning about configuration.  This section tries to give the bigger picture by showing how the pieces fit together and thus get people going quickly.

This is an initial draft that I'm still refining, still some missing bits.  For now, general feedback on the organization and the advice it gives to users would be most useful (also Should this new material in `utilities.rst` be a new chapter or go elsewhere? and Is this helpful?)

@Zimmi48 I'm looking at you.  Hope you can give some initial comments soon.

- [x] Added **changelog**.

Alizter commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,294 @@
+
+The mechanisms for creating logical paths for your multi-file project are:
+
+- Through command line options passed on the command line to `coqc` and `coqide`,
+  principally :n:`-R <filesystem path> <logical name>`.  The resulting logical path
+  is the logical name followed by a suffix of the filesystem path (as for `Coq`
+  above).  Use these options directly if you must.  Command line options are described
+  :ref:`here <command-line-options>`.
```
There is a subtle difference in preference with `-R` and `-Q` options.

When working on your own Coq project, I would say it is advisable to use `-R`. So if I had a `theories/` directory in my project, I would have the option `-R theories/ MyProject`.

When referencing another project in some directory that I want to use in my project I would prefer `-Q`. So if `OtherProject` lived in `path/to/other/` I would pass `-Q path/to/other/ OtherProject` to depend on it in MyProject.

The reason this is preferred is to not pollute the namespace. For example, OtherProject could have many files, many of which would be completely unrelated to MyProject. If I were to pass it as `-R` it would carry around a lot of needless baggage.

It might seem surprising that `-Q` is even enough here however. For example, if OtherProject had a deep folder structure, you might think that we need to specify all those subfolders too when depending on it. This is where some subtle Coq Require semantics step in. When you do `From OtherProject Require Import blah.`, Coq is actually able to search in subdirectories even if they were not specified! This means that coqdep (which is the workhorse finding dependencies) knows exactly where to find things.

This is the thing that Hugo worked on recently, which was to pin down precisely the semantics of this directory search when using `From`.

And finally, for the question of "Why have `-R` at all then?", the answer is that many Coq users do not like having to write `From MyProject Require Import myblah.` whilst inside MyProject, which is understandable.

jfehrle replied:
An overview of configuration won't describe every possible way to configure Coq.  We probably have more overlapping configuration knobs than we really need.  If `-Q` is useful for users I will mention it, but consider the following:

In the long term, it would be simpler if -Q -R and -I were not necessary, instead identifying packages solely by a directory name (including all subdirectories) and taking the full logical pathname from directory names.  That would be simpler for users.  So I wouldn't encourage the use of -Q consistent with this idea.

> When referencing another project in some directory that I want to use in my project I would prefer `-Q`. So if `OtherProject` lived in `path/to/other/` I would pass `-Q path/to/other/ OtherProject` to depend on it in MyProject.

That makes the _CoqProject system dependent: `path/to/other` will be not be in the directory containing _CoqProject or a subdirectory.  It may be an absolute path.  I thought that was not best practice, though it may be an easy shortcut that won't scale easily if you have multiple packages in your project (packages you maintain)--you likely would need to add the -Q entries into multiple `_CoqProject` files.

> The reason this is preferred is to not pollute the namespace.

"namespace" meaning the loadpath (logical names and directories) and when searching for files with `Require`.  I don't see how having more things in the loadpath is much of an issue, though `Print LoadPath` should probably have an option to show only the logical name.

As for the searching, I've been experimenting with creating my own `Bool.v`.  I can load my own IIRC and there's no indication that the builtin one has been hidden.  Users should have a way to report this.  The expected behavior is not documented.  This seems like a significant usability issue.

> This is the thing that Hugo worked on recently, which was to pin down precisely the semantics of this directory search when using `From`.

I remember that, but I've not been able trigger the "Required library qualid matches several files in path (found file1.vo, file2.vo, ...)." message.

> When you do `From OtherProject Require Import blah.`, Coq is actually able to search in subdirectories even if they were not specified!

Could you give a concrete example?

> And finally, for the question of "Why have `-R` at all then?", the answer is that many Coq users do not like having to write `From MyProject Require Import myblah.` whilst inside MyProject, which is understandable.

I think this reflects a weakness of how we currently resolve references.  Each package should be able to have local internal references without having to specify `From MyProject`, but reference resolution always uses the entire loadpath, i.e. global.  It's not so modular.  On the other hand, users probably don't want to have to write `From Coq Require Import Nat`, either.  I found 5800 uses of `Require` in the source tree but only 473 of them use `From`.

I'd like focus on adding useful info to the doc using what we have now.  I've found perhaps a dozen things that we could improve in configuration, but let's not get too diverting in analyzing those in depth.  I think the doc needs to clearly explain how conflicting/ambiguous references are handled.

Let me know your thoughts.  (The examples are invaluable, too.)

Alizter replied:
>In the long term, it would be simpler if -Q -R and -I were not necessary, instead identifying packages solely by a directory name (including all subdirectories) and taking the full logical pathname from directory names. That would be simpler for users. So I wouldn't encourage the use of -Q consistent with this idea.

I believe that it is simpler to only think of -Q and -I when it comes to Coq projects, since internally this is how it is handled. `-R` is simply a shortcut to recursively do -Q for each subfolder if I am not mistaken.

>That makes the _CoqProject system dependent: path/to/other will be not be in the directory containing _CoqProject or a subdirectory. It may be an absolute path. I thought that was not best practice, though it may be an easy shortcut that won't scale easily if you have multiple packages in your project (packages you maintain)--you likely would need to add the -Q entries into multiple _CoqProject files.

`From ... Require Import ...` should be able to find external libraries (as long as lib and bin have been setup correctly) without the need to pass -Q. Typically the only reason to pass -Q are for folders in the very project you are working on, or for referencing projects which are not installed globally. In both cases, a relative path is typically easy to come by. Absolute paths for global install directories should not need to be given, since Coq should be able to find those itself.

>"namespace" meaning the loadpath (logical names and directories) and when searching for files with Require. I don't see how having more things in the loadpath is much of an issue, though Print LoadPath should probably have an option to show only the logical name.

If you are using `From ... Require Import ...` syntax, then the loadpath having many entries should typically not pose much of an issue. I believe that if you are using `Require Import ...` then Coq will have to start searching through all the loadpaths (consistently between coqc and coqdep at least). Your later example of Bool.v could demonstrate this problem. If I had many projects with a Bool.v in my loadpaths, how would I know which one was imported by `Require Import Bool.`. The answer I believe is, which ever came first in the list of load paths, but that is not something the user should be expected to remember.

>As for the searching, I've been experimenting with creating my own Bool.v. I can load my own IIRC and there's no indication that the builtin one has been hidden. Users should have a way to report this. The expected behavior is not documented. This seems like a significant usability issue.

It depends on how you require/import Bool. Inside your project, you should do `Require Import Bool` and expect your local version to be loaded. But if for whatever reason you need the stdlib version, then you can do `From Coq Require Import Bool`. That is not ambiguous for Coq to find.

>> When you do From OtherProject Require Import blah., Coq is actually able to search in subdirectories even if they were not specified!

>Could you give a concrete example?

For instance, if I installed [coq-equations](https://github.com/mattam82/Coq-Equations) globally using its `make install` or using opam (`opam install coq-equations`), then I should be able to open a new .v file and do `From Equations Require Import Prop.Classes`. At no point did I have to specify where coq-equations was actually installed, that's what I mean by `From` being able to search in directories not specified.

Another example that doesn't require an external package is if I had a directory structure like:
```
.
└── A
    ├── A.v
    └── B
        └── B.v
```

And I had only `-Q A A` in my _CoqProject, then if I did `From A Require Import B`, Coq would be able to find B.v. This is what I meant by not having to specify subdirectories. Of course, `Require Import A.B` would not work, because it is not in the loadpath. And `-R A A` would have done something equivalent to `-Q A A`, `-Q A/B A.V` in the background.

SkySkimmer replied:
>Of course, Require Import A.B would not work, because it is not in the loadpath

I don't know what you mean by "not in the loadpath", it doesn't work because its name is A.B.B not A.B and Require doesn't insert components in the middle of qualids

Alizter replied:
Right my mistake, I meant `Require Import A.B.B` would not work. The module for the file A/B/B.v would not be seen by Coq correct?

SkySkimmer replied:
It would work. -Q and -R are the same except Require without From with -Q needs more explicit qualids.

Alizter replied:
OK I stand corrected. At least it's good that we are attempting to write this documentation. It still confuses me. 0:-)

Alizter reviewed (commented):
Good work on starting this! I have some initial comments.

proux01 commented on doc/sphinx/proof-engine/vernacular-commands.rst:
```diff
@@ -736,7 +737,7 @@ Extra Dependencies
 follows:
 
-.. cmd:: From @dirpath Extra Dependency @string {? as @ident }
+.. cmd:: From @qualid Extra Dependency @string {? as @ident }
```
I guess a rebase should fix it.

Alizter replied:
The rebase won't fix anything here since the change has been done.

proux01 replied:
It will undo it if doc_gram_rsts is run.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,294 @@
+the :n:`From @dirpath` option of :cmd:`Require` to make the behavior more predictable.
+
+
+Upgrading to a new version of Coq
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`.vo` files are specific to the version of Coq that compiled it.  When you
+upgrade to a new version of Coq, you must recompile all the projects
+that you want to run in the new version.  This is necessary to assure that
+your proofs still work in the new version.  Once their projects build on the
+new version, most users no longer have a need to run on the old version.
+
+If, however, you want to overlap working on your project on both the old and new
+versions, you'll need to create separate source directories for your project
+for the different Coq versions.  Currently the compiled `.vo` files are kept
+in the same directory as their corresponding `.v` file and the `.vo` is specific
+to one version of Coq.
```
There are two ways to work with multiple versions: you can separate them by space or time.   "Space" meaning by the use of multiple copies in your file system, and "time" meaning you have a single copy that you update to the particular version you need at the moment.  Now that you mention it, IIRC one a switch is created, opam doesn't need to recompile it, right?  That would work neatly for any Coq package that was installed with opam.

But how about the code for a project that someone is actively developing?  You can make that a local opam repository, add that as an additional repository, configure the switch and then actually switch.    (Do they have to have it in git?  What if it's not in github?)  If you know of an existing explanation of how to do this or you can provide a short but reasonably complete explanation of how to do that, I can try to work that in.  Or I can just suggest the possibility without much detail.

Alizter replied:
You don't have to make your own projects into opam projects to use Coq and it's packages installed in an opam switch. They will look like they were installed globally as usual. As long as your path variable is configured to point to the opam bin and lib folders, everything should work as with a usual install.

The pointing to lib and bin is exactly what the command `eval $(opam env)` does. Although I don't know what the windows equivalent for this would be. It should definitely work in cygwin AFAIK however.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,294 @@
+
+The components of each pair share suffixes, e.g. "Bignums.BigZ" and "Bignums/BigZ" or
+"Coq.Numbers.Natural" and "Numbers/Natural".  Physical pathnames should
+always use `/` rather than `\\`, even when running in on Windows.
```
No, it's `coq_makefile`, which reports an error if you have `\` in the _CoqProject.  Though you're right, cygwin understands backslash in paths, but they have to be quoted in the command line.

Zimmi48 commented:
This won't be ready for the branching date, but since this is only doc, maybe this could still be considered for inclusion if ready after this date and before the final release?

ppedrot commented:
Let's postpone for now and see later.

brando90 commented:
> Let's postpone for now and see later.

When will this be ready?

fyi: CoqDicuss link related to this: https://coq.discourse.group/t/official-place-to-learn-how-to-setup-coq-make-files-for-beginner/1682/2

jfehrle commented:
@brando90 It's hard to predict.  I hope this is finished within the next few weeks, but it may take take longer and sometimes reviews take a very long time.  Everyone is a volunteer.

This [link](https://coq.gitlab.io/-/coq/-/jobs/2320802692/artifacts/_install_ci/share/doc/coq/sphinx/html/practical-tools/utilities.html#coq-configuration-basics) will show you the formatted manual with the changes.  That link will eventually stop working.  You can also access it from here in this page:

![image](https://user-images.githubusercontent.com/1253341/173129211-1e5f09ec-ddbf-42b5-adfe-1b729ac7946b.png)

Speaking of which, @Alizter would you care to complete the review and be the assignee sometime soon?

Alizter reviewed (commented):
Sorry about that. I forgot to submit the review from 2 days ago.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+same time.  While there's currently no guarantee that user-developed packages
+will compile on the current version of Coq, all packages
+that Coq platform installs should compile without difficulty--this is part of
+the Coq platform release process.  When you upgrade to a new version of
+Coq platform, the packages it installed will be updated to as needed to
+remain consistent.
```
Consistent with each other.  I'll drop the sentence, too much detail and sort of obvious.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+- If you installed with opam, use :n:`opam show --list-files coq-bignums | head -n1` -
+  the last component of the filename is the logical name (`Bignums`).
+
+- On Linux, :n:`ls $(coqtop -where)/user-contrib` shows the logical names of all
+  installed user-contributed packages.  You should be able to guess which one you
+  need.
```
Try it.  Prove me wrong if you can.  The directory structure in `user-contrib` exactly matches the module structure.  The source code may have directories like `src` that don't match, but the installation process would have to move it.  _CoqProject files in user-contrib have no effect on loading modules.  Directories there are always loaded as if they had `-Q <dirname>` regardless of _CoqProject.

> Could you put that higher up?

The first bullet method always works with no guessing.  Second bullet may require some easy guessing.  Output of `Print LoadPath` can be very long and only a few lines are relevant.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+Setup for working on your own projects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The working and master copies of source code for your own projects should
+be kept separate from the Coq installation.  In particular, when you upgrade
+to a new version of Coq, any directories you created in the old version won't
+be copied or moved.
```
Yes, it would be strange.  Just trying to keep beginners out of trouble.

I made it: "The working and master copies of source code for your own projects should
not be in the directory tree where Coq is installed."

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+  File1.v
+  SubDir/File2.v
+
+TODO PR: Should we recommend -R (easier but more likely to give ambiguous paths) or
+-Q (safer, less likely to give ambiguous paths)?
```
Yeah.  I'm learning and figuring out lots of things as I write this.

I think `-Q` is not ideal for referring to an external project because the associated directory is system-dependent.  (It will certainly be incorrect if publish your development in opam.)  You don't need it and shouldn't use it to refer to something that's installed, which shouldn't be re-compiled.  It would be good to have another flag to identify what packages the project depends on that wouldn't affect compilation, e.g. `-D Bignums` for `coq-bignums` or maybe `-D Bignums coq-bignums`.  For the future.

Alizter replied:
The reality is that today, you do not need to set -Q and -R flags for other projects. Today, I can install mathcomp via opam and start using it in my own Coq project. I won't need to set -Q or -R or know anything about how mathcomp is installed. All I have to do is write `From mathcomp Require Import ...`, Coq will be able to find the file I specify no problem.

Internally, Coq knows that user-contrib is where it should start looking for things. Every directory in user-contrib gets `-Q` set for it automatically. That way the Coq command `From my_installed_dep Require someFile` can always be resolved.

Alizter replied:
Don't worry about having to learn all these details when writing these docs. The details are awfully complicated and I am still learning a lot of new things every time I look into the details. The fact that we are even attempting to document this behaviour is a huge step in the right direction.

jfehrle replied:
> The reality is that today, you do not need to set -Q and -R flags for other projects.

Not needed for installed projects such as mathcomp, not needed if you use COQPATH for packages outside of the installation directory.  Maybe a useful shortcut if you don't use COQPATH.  Not that I love COQPATH...

> Don't worry about having to learn all these details when writing these docs.

It's hard write anything useful without considerable understanding of the material.  I write stuff, then figure out that it not right, then I redo it.  Sometime multiple iterations.  As long as it's right in the end (or at least clear enough to be correctible).

Alizter replied:
I meant that: Of course you need to know what is going on to write about it, but don't worry about not understanding what is going on since there details are complex and most of us barely understand it anyway.

Alizter replied:
Sure you could use COQPATH to also direct Coq to find the compiled vo files, however this is more of an "installation thing". If you are actively developing a multi-subproject project, you probably don't want to be "installing" parts of it for it to work.

Alizter replied:
@SkySkimmer Would you mind correcting me then?

SkySkimmer replied:
-R is like -Q but allows require without an explicit prefix

Alizter commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+you like and run Coq from that directory.  If your project has multiple files,
+then you need to compile them in an order that matches their dependencies.
+If B.v depends on A.v, then you should compile A.v before B.v.  You can do this
+with :n:`coqc A.v B.v`, but you may find it tedious to identify the dependencies.
```
COQPATH is not really needed by anybody. It simply allows for people to install libraries in places other than user-contrib. If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.

The `Cd` commands in Coq are really ready to be deprecated any day. They unfortunately have one valid use case with extraction today, but this is really a lack of a feature in extraction, than a valid use of Cd. We should refrain as much as possible from mentioning them.

Even a multi-directory project, will not require setting COQPATH. The only reason you might want to use COQPATH is if you have two separate Coq projects which you don't want to install, but you would like to use one from the other as if it has been installed. However, I still think explaining this to a starting user is pointless, and it should be a technical packaging detail.

jfehrle replied:
Let's go through this a bit:

> If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.

That was not what I noticed when I tried this two months ago, though I was working from my own build out of the source tree.  At the moment I find `coq_makefile` won't even run for me with a fresh 8.15.2 installed with opam.  See #16183.  What is your command line to make for the coq_makefile-generated makefile?

(How exactly can I build and test coq_makefile locally when there's no dune target for it?  I need to do something like a `dune install`, then run the installed version.)

IIRC, building with coq_makefile left the .vo files in the same directory as the .v file.  You certainly don't want to put the source code you're busily modifying into the Coq installation directory, let alone its git files.

On general principles, I think it's a poor practice to copy package(s) you're in the midst of modifying into the official installation directories--it's too easy to mess something up even if its scripted.  The installation directories should changed only by opam or when you explicitly install a package.

For example, suppose you use bignums for your work and you also happen to be a maintainer of bignums.  It will be all too easy to end up with the wrong binaries for bignums in the installation directory.  If you upgrade Coq with Coq platform, the modified files will, in effect, disappear because it will be in a new switch.  Not sure what would happen with opam if you change the coq version or a package version directly.

The IDEs will find the _CoqProject associated with a script file based on the file path, but that won't work if a) you try to access the package from an unnamed buffer or b) from a file that's outside of your package.

Let me know your thoughts.  There are some facts to check, maybe you can see if `coq_makefile` and the file it generates works for you.  Two plus months after I wrote this, it takes a while to reconstruct my reasoning.

> The `Cd` commands in Coq are really ready to be deprecated any day. 

Yeah, I saw that.  It may be useful as IDE-only concept rather than as a command.  The doc says "The `<>` in the final entry represents an empty logical pathname, which permits loading files from the associated directory with just the basename of the script file, e.g. specify `Foo` to load `Foo.vo`.  Note that the :cmd:`Cd` command doesn't change the associated directory--you would need to restart CoqIDE."

Alizter replied:
> Let's go through this a bit:
> 
> > If you have a coq_makefile project or dune project, they will both install into user-contrib today by default.
> 
> That was not what I noticed when I tried this two months ago, though I was working from my own build out of the source tree. At the moment I find `coq_makefile` won't even run for me with a fresh 8.15.2 installed with opam. See #16183. What is your command line to `make` for the coq_makefile-generated makefile?

If I generate a makefile with `coq_makefile -f _CoqProject -o Makefile` I will get a makefile called `Makefile`. If I then run `make` in the same directory then the project will be built. To install it, the makefile will contain an `install` target which I can run using `make install`. This will the proceed to install in the install directory.

If you installed Coq via opam, then you will have likely come across a command such as `eval $(opam env)`. This is very important. If you run `opam env` you will see it output some text about setting some environment variables. The `eval $(..)` part will set those variables for your terminal session. If you set up opam (during opam init) to configure something in your user profile, then this will automatically be run on login.

The important environment variable that gets set here is OCAMLPATH. This is the install location for OCaml libraries. Coq is an OCaml library so it can be found in `$OCAMLPATH/[lib/?]Coq`. (I'm not sure about the lib part). Inside the Coq directory there is the user-contrib directory in which Coq libraries themselves get installed.

So how does Coq know about the install directory (in coq_makefile)? It first checks OCAMLPATH for the Coq folder. After which it has access to theories/ and user-contrib/. It then installs into the second.

The Coq folder in OCAMLPATH can be overridden by specifying COQLIB. Which for the most part is unset. If it is set Coq will take precedence for that over the guessing game above.

> (How exactly can I build and test coq_makefile locally when there's no dune target for it? Do I need to do something like a `dune install`, then run the installed version?)

I think there might be some confusion between the actual Coq targets, and the shims. Let me try to clear up some of this.

As I described above, Coq when starting, does a lot of guesswork trying to understand where the Coq stdlib might be and where the installation locations are. Part of its validation for the guess of COQLIB is to look for a very specific file in the stdlib called `theories/Init/Prelude.vo`.

This is the "Coq prelude" which is implicitly (`Require`d) in every Coq session. Therefore Coq needs to know where it is in order to use it. You can disable this implicit loading with the `-noinit` option.

So why do we need shims? Well if you load a Coq binary, without the prelude file present, you will get a message saying something like "coqlib not found". This is kind of strange, since coq_makefile doesn't even do any proof assisting, it is just a tool to generate a makefile. The answer to why this is, is that most of the Coq binaries share the same `boot/env.ml` file which takes care of setting up the COQLIB business I described above.

This is not an ideal factorization, since it means that coq_makefile has to do some unrelated things. But it historically allowed for sharing functions such as printing command line help options etc.

Back to the shims: The `dev/shim/` folder's job is to make sure enough of the Coq prelude is built, so that the binaries do not complain. That is all that it does. To get the same effect as say `dune build dev/shim/coqc-prelude` you need to build both the prelude and the `coqc` binary. This can be done first by writing `dune build theories/Init/Prelude.vo`, Dune will then go and starting building all the necessary things and eventually that.

Next you want to build the `coqc` binary. If you try `dune build coqc` you will see that it does not work. The reason is that there is no build target named `coqc` in the Coq directory. What there is instead is an executable called `coqc` which will live somewhere in the Coq source tree. It is not important to locate these executables, Dune can do it for you. This is what the `dune exec -- coqc` command is doing. The `--` just means "parse the rest as a single command line arg", this makes passing args to coqc easy.

`dune exec` first goes and looks for an executable called `coqc`, which is declared in https://github.com/coq/coq/blob/master/topbin/dune. It then tries to build it and run it.

All of this is exactly what the `dev/shim/coqc-prelude` shim is doing, but with some environment state also being set.

```lisp
(rule
 (targets coqc-prelude)
 (deps
  %{bin:coqc}
  %{project_root}/theories/Init/Prelude.vo)
 (action
  (with-stdout-to coqc-prelude
   (progn
    (echo "#!/usr/bin/env bash\n")
    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqc} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
    (run chmod +x %{targets})))))
```

You can see that the `(rule )` stanza declares `%{bin:coqc}` (the coqc binary), and `Prelude.vo` as dependencies. This means these need to get built before you do what is described in the `(action )` field. All that is happening in the action field however is `coqc` is being run, with an explicit `-coqlib` and (`-I` for coq plugins) being set to the ones that Dune just built. This allows you to run your executable in the correct env.

Very confusingly, it is actually creating a wrapper bash script around coqc with the env variables in that. (The with-stdout-to part pipes the output (of echo) to a file named `coqc-prelude`. The `(progn )` part runs several command in succession. You can even see the `chmod` part letting you run the script at the end.

Let's come back to coq_makefile. How can we create a shim for that? Well it's probably very easy, just replace mention of coqc in the above thing with `coq_makefile`. You also probably don't need the `-I` part but I don't think it does any harm. `coq_makefile` also requires some extra files such as `tools/Makefile.in` IIRC, so you need to add that to the deps field. In Dune a good rule of thumb is to always declare your deps, even if you know they will always be there.

> IIRC, building with coq_makefile left the .vo files in the same directory as the .v file. You certainly don't want to put the source code you're busily modifying into the Coq installation directory, let alone its git files.

This is how makefile builds work unfortunately. They will produce the .vo targets in the same directory as the .v files. However, the installation part is exactly the job of `make install` which will systematically copy and install the `.vo` files to the installation directory it determines in the mess above.

> On general principles, I think it's a poor practice to copy package(s) you're in the midst of modifying into the official installation directories--it's too easy to mess something up even if its scripted. The installation directories should changed only by opam or when you explicitly install a package.

Absolutely, nothing should have to be installed when developing a package.

> For example, suppose you use bignums for your work and you also happen to be a maintainer of bignums. It will be all too easy to end up with the wrong binaries for bignums in the installation directory. If you upgrade Coq with Coq platform, the modified files will, in effect, disappear because it will be in a new switch. Not sure what would happen with opam if you change the coq version or a package version directly.

Typically, if you are maintaining the packages in Coq platform, then you probably should not be using Coq platform to do this. Coq platform is supposed to be an easy-to-install multi-platform solution for getting Coq working, rather than a development environment.

Something that might be of interest is the recent inter-project composition for Coq feature that Rudi and I worked on in Dune. This let's you say work on bugnums and another project that depends on it simultaneously, since Dune now knows how to "compose" and build these projects in tandem. Getting smooth IDE support for this is something that needs to be worked on.

> The IDEs will find the _CoqProject associated with a script file based on the file path, but that won't work if a) you try to access the package from an unnamed buffer or b) from a file that's outside of your package.

I can't speak for PG, but coqidetop based IDEs (CoqIDE and VSCoq) have some trouble with _CoqProject files. It only really works when you open a named buffer in the scope of a _CoqProject file. Even if you open a new buffer and save it, I find myself having to reopen before the _CoqProject is properly detected.

For multi-project developments, that is all too true. CoqIDE really doesn't work that well. One solution is to have two instances of CoqIDE running, per _CoqProject file. This can typically have better results.

> Let me know your thoughts. There are some facts to check, maybe you can see if `coq_makefile` and the file it generates works for you. Two plus months after I wrote this, it takes a while to reconstruct my reasoning.

I've posted a lot of information in this reply, so hopefully it is useful. I think if we iterate on this, we will have much better documentation than before.

> > The `Cd` commands in Coq are really ready to be deprecated any day.
> 
> Yeah, I saw that. It may be useful as IDE-only concept rather than as a command. The doc says "The `<>` in the final entry represents an empty logical pathname, which permits loading files from the associated directory with just the basename of the script file, e.g. specify `Foo` to load `Foo.vo`. Note that the :cmd:`Cd` command doesn't change the associated directory--you would need to restart CoqIDE."

The `Cd` command is really an archaic artefact from the time people used `coqtop` to write Coq files. It is terrible for `coqc` for example. It actually changes where the `.vo` file will get ouput!

jfehrle replied:
Thanks for the detailed response.  Perhaps a bit too much detail.  I hoped that one of you experts would eventually create a `coq_makefile` shim to bring dune closer to parity with `make`-based builds.

The main point is that given the behavior of the makefile generated by `coq_makefile`, users need to know about COQPATH.  Agreed?

> Typically, if you are maintaining the packages in Coq platform, then you probably should not be using Coq platform to do this. Coq platform is supposed to be an easy-to-install multi-platform solution for getting Coq working, rather than a development environment.

Huh?  If your package only has .v files and no OCaml, you don't need to have the Coq source tree.  

> Something that might be of interest is the recent inter-project composition for Coq feature that Rudi and I worked on in Dune.

An interesting use case, but let's discuss after this PR is completed.  May need some tweaks to Coq.

> coqidetop based IDEs (CoqIDE and VSCoq) have some trouble with _CoqProject files. It only really works when you open a named buffer in the scope of a _CoqProject file. Even if you open a new buffer and save it, I find myself having to reopen before the _CoqProject is properly detected.

The _CoqProject parameters become CLI arguments to coqtopide.  The only way to change them is to start a new coqtopide.  Resetting the buffer is sufficient and less annoying than closing and reopening.  This can be improved, CoqIDE should be checking for modified files.  Also you ought to be able to edit _CoqProject inside CoqIDE in a buffer.

> For multi-project developments, ... CoqIDE really doesn't work that well. One solution is to have two instances of CoqIDE running, per _CoqProject file. This can typically have better results.

If you run two instances, changes made in instance 1 are not visible in instance 2 until you reset the buffer and create a new coqtopide.  This can also be improved.

JasonGross replied:
> COQPATH is not really needed by anybody.

COQPATH is absolutely essential in Fiat Crypto and it's dependencies, which can be configured to use either installed dependencies or to use submodules for dependencies.  We set COQPATH to make use of submodule dependencies, which are often built with different flags than the main project.

jfehrle replied:
Agreed, COQPATH is needed.  I recommend its use in the utilities chapter.  @Alizter, if you still disagree, please speak up.

Alizter replied:
What I meant is that it is not essential to use COQPATH in order to load installed libraries. I did not intend to sound like it was unused.

SkySkimmer commented on vernac/vernacentries.ml:
```diff
@@ -483,7 +483,7 @@ exception NotFoundLibrary of Names.DirPath.t option * Libnames.qualid
 
 let _ = CErrors.register_handler begin function
-  | UnmappedLibrary (from, qid) -> Some (err_unmapped_library ?from qid)
+  | UnmappedLibrary (from, qid) -> Some (err_unmapped_library ?from qid) (* error msg*)
```
??

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -7,32 +7,307 @@
+TODO PR: Should we recommend -R (easier but more likely to give ambiguous paths) or
+-Q (safer, less likely to give ambiguous paths)?
+
+TODO: Are -R/-Q with dotted names useful?  If so, need to explain and adjust the
+description for :term:`logical name`.  For example, this doesn't produce usable
+.vo files because a.v and b.v are not in an `Abc` subdirectory::
```
Not today.  `a.vo` will be in the current directory, but `Michael.Abc` will look for `Michael/Abc/a.vo`.  So I think not useful.

jfehrle commented:
Updated.  Let me know if you have more comments.  I'll probably have more changes of my own.

Still need to think about where to put this section.  This is not an obvious place for it:

![image](https://user-images.githubusercontent.com/1253341/173658623-3b8db83a-576a-499d-8368-10ef8145b7f9.png)

Also the introduction to "Building a Coq project" needs some changes so the new section fits smoothly with it.

jfehrle commented:
@Alizter, let me work on it some more before you go through this again.

jfehrle commented:
Updated.  I made a point of describing the -R/-Q options (coq-commands chapter), logical names (modules chapter) and the load path (utilities chapter) each in a single place, removing repetitive descriptions that said different things (i.e.  confusing) , and adding appropriate hyperlinks.  Also I moved the utilities chapter in the TOC and renamed it "Building Coq Projects".

There's still more to improve in this vicinity, but I'll save that for future PRs.

@Alizter would you take another look?  I'll squash before merging.

SkySkimmer commented on doc/sphinx/language/core/modules.rst:
```diff
@@ -879,81 +879,71 @@ and :math:`Γ_C` is :math:`[c_1{:}∀ Γ_P, C_1 ; …; c_n{:}∀ Γ_P, C_n ]`.
+as :n:`Coq.Init.Logic`, in which:
+
+- :n:`@ident__logical`, the :gdef:`logical name`, maps to a specific
+  directory (or :gdef:`physical path`) in the user's file system.  The logical name
```
Note it is possible to associate a logical path to multiple physical paths

jfehrle replied:
Got it.  I wonder if multiple physical paths is a good practice (or merely necessary at times).

jfehrle replied:
Also, I didn't mention possibilities such as `-R . Foo.Bar`, thinking they are not very useful.  (My impression is that we've made some of the config details a bit too general, giving lots of corner cases that are not that useful.  I've tried not to dwell on these, they mostly add complexity/confusion.)

SkySkimmer commented on doc/sphinx/practical-tools/coq-commands.rst:
```diff
@@ -148,39 +169,53 @@ except that ``space_overhead`` is set to 120 and ``minor_heap_size`` is set to 3
+  case.
+
+  Loading files from packages made available with `-Q` must include
+  the :term:`logical name` of the package in `From` clause of the :cmd:`Require`
```
or the Require may use the fully qualified name, eg `Require Ltac2.Ltac2`

jfehrle replied:
Also, I didn't mention possibilities such as `-R . Foo.Bar`, thinking they are not very useful.  (My impression is that we've made some of the config details a bit too general, giving lots of corner cases that are not that useful.  I've tried not to dwell on these, they mostly add complexity/confusion.)

SkySkimmer commented on doc/sphinx/proof-engine/vernacular-commands.rst:
```diff
@@ -558,17 +572,18 @@ file is a particular case of a module called a *library file*.
+   made available as if with a `-Q` option.  If you create a project outside of these
+   using `-R` in `_CoqProject` and then later put in one of these locations,
+   :cmd:`Require` commands in your code that depend on the `-R` behavior will fail in
+   the new location.  Therefore we recommend using only `-Q` and not `-R`.  In contrast,
```
Do we really recommend that?

Zimmi48 replied:
I agree that we don't recommend that. In fact, moving one part of your project to `user-contrib` and keeping the rest somewhere else is what should be discouraged. Only use `-R` to refer to what's inside your library, that you will install at once in `user-contrib`.

jfehrle replied:
I missed updating this.  I'll say something similar to @Zimmi48's wording, e.g. "Only use `-R` to refer to files within a single package.  Use `-Q` to refer to files in other packages".

Also, the following may also be incorrect--do you guys know?  Are partially qualified names resolved at compile time or run time?  My guess is that it's at compile time (in which case this text is ~100% incorrect); resolving PQ names at run time seems like it could lead to obscure errors.

>    Also note that packages installed in `user-contrib` or accessed through `COQPATH` are
   made available as if with a `-Q` option.  If you create a project outside of these
   using `-R` in `_CoqProject` and then later put in one of these locations,
   :cmd:`Require` commands in your code that depend on the `-R` behavior will fail in
   the new location.

SkySkimmer replied:
>Also, the following may also be incorrect--do you guys know? Are partially qualified names resolved at compile time or run time?

I'm not sure what you mean by "run time" (my default interpretation would be the same as "compile time"). If you're asking what's in .vo files then indeed they're fully qualified.

jfehrle replied:
Perhaps I should have said "load time".  Just carrying over programming language terminology imprecisely.

jfehrle commented:
Updated (limited changes).

Alizter commented:
@jfehrle Just to let you know, I am in the middle of reviewing this.

Alizter reviewed (approved):
LGTM. This is a vast improvement from before. I am happy with merging this soon.

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -1,53 +1,371 @@
+.. todo I thought @herbelin added code to complain about ambiguous short names
+   I made up some stuff below, need to check it:
+
+If you associate the same logical name with more than one directory, Coq will
+use the last one added (i.e., that appears earlier in :cmd:`Print LoadPath`).
+Nor will Coq complain if the short name of a package is ambiguous.  Perhaps
+in the future, Coq will be able to report such conflicts.  We recommend using
+the :n:`From @dirpath` option of :cmd:`Require` to make the behavior more predictable.
```
It reports an ambiguity for `Require` but not for `From ... Require`.  See #16314.  When I do the equivalent with `_CoqProject`, there's no visible error, either:

```
$ more _CoqProject
-R . Ambig
a.v
ambig2/a.v
$ make -f CoqMakefile
COQC a.v
COQC ambig2/a.v
$ coqc --version
The Coq Proof Assistant, version 8.15.2
```

I will reword a bit.

Alizter replied:
As I explained in the issue, that is not really ambiguous. What would be ambiguous for both `From ... Require` and `Require` would be:
```
A/B/a.v
A/C/a.v
```

and you wrote one of:
```
Require a.
From A Require a.
```

jfehrle replied:
> The actual specification of what is considered ambiguous however is very complicated and not fully written out.

Not written out at all.  At the very least, the `Require` command should describe the behavior so users understand it.  The `Require` doc doesn't mention "shortest" anywhere so the behavior contradicts the description.  The complexity may be justifiable for compatibility reasons and to avoid the need for pervasive changes to `Require` commands.  If someone can describe the behavior soon I'll include that in this PR.

Alizter replied:
@jfehrle I don't think it is a simple task describing the behaviour of the `From ... Require` command. It has a very subtle semantics. We can paint a picture of some of the important parts with some broad strokes, but a full description of how the algorithm works will be difficult to compile.

Historically, the technical difficulty and subtlety of this algorithm has been the main barrier to its documentation. Hugo tried to at least coincide the behaviours of coqdep and coqc, together with writing many tests demonstrating the expected behaviour. I believe it would be an entire project onto itself to describe with further details what is going on. Not something I am against starting however.

jfehrle replied:
I removed lines 342-344.  I understand it's an exact match, which would be a shortest match, but "shortest" doesn't apply in general.

jfehrle commented:
Updated.  Please take a quick look at the final changes in the last commit, then I'll squash.

Alizter commented:
Looks okay, please squash.
